src/stockpile.clj

```diff
@@ -0,0 +1,364 @@
+    (durably-establish (.resolve top "stockpile")
+                       ;; Assumes that copy won't use a
+                       ;; BufferedWriter (FilterOutputStream) in this
+                       ;; case (otherwise it'll be broken with at
+                       ;; least openjdk-7.
+                       #(with-open [out (FileOutputStream. (.toFile %))]
+                          (.write out (.getBytes "0 stockpile" "UTF-8")))
+                       false)
+    (fsync top true)
+    {:directory top
+     :encode-meta encode-meta
+     :next-likely-id (AtomicLong. 0)}))
+
+(defn open
+  "Opens the queue in directory, and returns the queue.
+  Calls existing-entry-reducer as-per reduce for each existing entry
+  in q (the ordering of the calls is unspecified).  The reduction may
+  be escaped by throwing a unique exception (cf. slingshot).  Returns
+  [q reduction].  For example: (open \"foo\" conj []).  Calls
+  (encode-meta metadata-string) and (decode-meta encoded-string) to
+  handle encoding and decoding entry metadata to and from filesystem
+  Path strings."
+  [directory existing-entry-reducer reduction-base-val
+   & {:keys [encode-meta decode-meta]
+      :or {encode-meta encode-portable-path-element
+           decode-meta decode-portable-path-element}}]
+  (let [top (as-path directory)
+        q (.resolve top "q")]
+    (let [info-file (.resolve top "stockpile")
+          info (String. (Files/readAllBytes info-file) "UTF-8")]
+      (when-not (= "0 stockpile" info)
+        (throw (IllegalStateException.
+                (format "Invalid queue token %s found in %s"
+                        (pr-str info)
+                        (pr-str info-file))))))
+    (let [[q reduction] (fs->queue top decode-meta
+                                   existing-entry-reducer reduction-base-val)]
+      [(assoc q :encode-meta encode-meta) reduction])))
+
+(defn store
+  "Atomically and durably enqueues the content of stream, and returns
+  an entry that can be used to refer to the content later.  If an
+  ex-info exception of {:kind ::commit-failed} is thrown, it's
+  possible that the metadata was too long for the underlying
+  filesystem.  If provided, (adjust-metadata retry-count
+  previous-metadata) will be called whenever there is a failure that
+  might be related to the length of the metadata string.  It can
+  either return an alternate string, or nil/false to give up, which
+  will cause the ::commit-failed exception."
+  ([q stream] (store q stream nil))
+  ([q stream metadata] (store q stream metadata nil))
+  ([q stream metadata adjust-metadata]
+   (let [likely-id (.getAndIncrement (:next-likely-id q))]
+     (when-not likely-id
+       (throw (IllegalStateException.
+               (format "cannot write to queue in %s while opening it"
+                       (pr-str (-> (:directory q) .toAbsolutePath str))))))
+     (let [tmp-dest (create-tmp-file (.resolve (:directory q) "q"))
+           encode-meta (:encode-meta q)]
```

Comment:
@mullr: @senior and I chatted about this, and I think I'm OK with removing the length-based retry/callback process.  I'd scared myself briefly last week when I misinterpreted some test results and believed that one of the situations we care about had a 255 *absolute* path-length limit.  It doesn't.

However, that still leaves the separate question of encodings and/or portability.  I'm not sure what we care about promising, but it looks like the JVM encodes Paths and Files according to the locale.  If so, then presumably (at least on systems where the paths are more or less just bytes - e.g. Linux ext4/btrfs/xfs), changing the locale could make an existing queue unreadable.  We'll need to decide if that's OK.

Here's an example you can run that hopefully demonstrates the underlying isuue.  i.e. that the path the JVM creates (and presumably reads - not demonstrated in this example) for a given string path depends on the locale:

```bash
#!/usr/bin/env bash

# Run this from an empty temp directory.

# Should work on at least Debian after making sure the two locales
# mentioned below have been enabled in "dpkg-reconfigure locales".
# You can see what locales you have available right now via "locale
# -a".

set -e

clojure_jar="${HOME}/.m2/repository/org/clojure/clojure/1.8.0/clojure-1.8.0.jar"

clojure()
{
  java -cp "$clojure_jar" clojure.main "$@"
}

create-sleep()
{
  clojure -e "(spit (str \"sue\" (char 16rf1) \\o) true)"
}

require-locale()
{
  if ! locale -a | grep -E "^$1$"; then
      echo 'Locale $1 does not appear to be available' 1>&2
      exit 1
  fi
}

require-locale es_MX.iso88591
require-locale en_US.utf8

# In Spanish, the "ñ" in "sueño" is "latin small letter n with tilde".
#iso8_sueno="$(echo -e 'sue\xf1o')"
#utf8_sueno="$(echo -e 'sue\xc3\xb1o')"

LC_ALL=es_MX.iso88591 create-sleep
LC_ALL=en_US.utf8 create-sleep

echo 'Now run "ls" in this directory from two different xterms:' 1>&2
echo '  LC_ALL=es_MX.iso88591 xterm -lc' 1>&2
echo '  LC_ALL=en_US.utf8     xterm -lc' 1>&2
```